pycheribuild/projects/cross/snmalloc.py

```diff
@@ -0,0 +1,87 @@
+# ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+# SUCH DAMAGE.
+#
+import shutil
+
+from .crosscompileproject import *
+from ...utils import getCompilerInfo, runCmd, IS_FREEBSD
+
+class SNMalloc(CrossCompileCMakeProject):
+    projectName = "snmalloc"
+    repository = GitRepository("https://github.com/Microsoft/snmalloc")
+    crossInstallDir = CrossInstallDir.CHERIBSD_ROOTFS
+    appendCheriBitsToBuildDir = True
+    supported_architectures = [CrossCompileTarget.CHERI, CrossCompileTarget.NATIVE, CrossCompileTarget.MIPS]
+    defaultOptimizationLevel = ["-O2"]
+    default_build_type = BuildType.DEBUG
+    defaultCMakeBuildType = "Debug"
+
+    dependencies = ["cheribsd-sysroot"]
+
+    @classmethod
+    def setupConfigOptions(cls, **kwargs):
+        super().setupConfigOptions(**kwargs)
+
+        cls.debug            = cls.addBoolOption("debug", help="Turn on debugging features")
+        cls.pagemap_pointers = cls.addBoolOption("pagemap-pointers", help="Change pagemap data structure to store pointers")
+        cls.pagemap_rederive = cls.addBoolOption("pagemap-rederive", help="Rederive internal pointers using the pagemap")
+        cls.cheri_bounds     = cls.addBoolOption("cheri-bounds", help="Set bounds on returned allocations")
+
+        # XXX Make this set the policy somehow?
+        cls.quarantine       = cls.addBoolOption("quarantine", help="Quarantine deallocations")
+        cls.revoke           = cls.addBoolOption("revoke", help="Revoke quarantine before reusing")
+
+    def __init__(self, config: CheriConfig, *args, **kwargs):
+        super().__init__(config, *args, **kwargs)
+
+        if self.revoke:
+            self.quarantine       = True
+            self.cheri_bounds     = True
+            self.pagemap_rederive = True
+
+        if self.cheri_bounds:
+            self.pagemap_rederive = True
+
+        if self.pagemap_rederive:
+            self.pagemap_pointers = True
+
+        self.COMMON_FLAGS.append("-DSNMALLOC_PAGEMAP_POINTERS=%d"   % (1 if self.pagemap_pointers else 0))
+        self.COMMON_FLAGS.append("-DSNMALLOC_PAGEMAP_REDERIVE=%d"   % (1 if self.pagemap_rederive else 0))
+        self.COMMON_FLAGS.append("-DSNMALLOC_CHERI_SETBOUNDS=%d"    % (1 if self.cheri_bounds     else 0))
+        self.COMMON_FLAGS.append("-DSNMALLOC_QUARANTINE_DEALLOC=%d" % (1 if self.quarantine       else 0))
+        self.COMMON_FLAGS.append("-DSNMALLOC_REVOKE_QUARANTINE=%d"  % (1 if self.revoke           else 0))
```

Comment:
Are there CMake options for these flags? If so I would use self.add_cmake_option() instead (which would handle boolean values)

Reply:
I hadn't added them, but I can.

Reply:
If they're real boolean values (i.e. you're not relying on `__bool__()` or `None`) then `%d` will do the coercion for you (and even if it didn't, `int(foo)` does).